juju/juju
# WIP: Suppress build number if the local jujud is an official build

## Description of change

When packaging a local jujud binary for bootstrapping we now check for a version file `jujud.ver` beside the binary. If the SHA256 of this jujud matches one in the file we consider this binary to be from an official build and don't tag the uploaded agent bundle with a build number.

This supports bootstrapping with a pre-downloaded binary to a restricted networking environment without the deployed controller being forced onto the devel stream.

The version file should look like this:
```yaml
versions:
  - version: 2.3-beta2-xenial-amd64
    sha256: b6813a18f82b16ae8d0cfb9e3063302688906e0c547db629a94dfb7f70198f00
  - version: 2.3-beta4-artful-arm64
    sha256: f7851b7a9bbd447944830fc9945010cfedf3b9ffdb39582ed14652747bd22546
...
```

It will be generated at packaging time and included alongside jujud.

## QA steps

Use this script to generate a version file for a `jujud` binary:
```bash
#!/bin/bash

jujud=$(which jujud)
version=$(jujud version)
hash=$(sha256sum $jujud | cut -d " " -f 1)
sigfile=$(dirname $jujud)/jujud.ver

rm $sigfile
cat > $sigfile <<EOF
versions:
  - version: $version
    sha256: $hash
EOF
```

Then bootstrap using the `juju` alongside that `jujud` - the version needs to be one that isn't in the streams yet. The controller should bootstrap without a .1 build number.

Killl the controller, delete the `jujud.ver` file and bootstrap again - the controller will have a .1 version.
